Source/objects.cpp

```diff
@@ -2728,37 +2728,40 @@ void OperateShrineReligious(Player &player)
 	InitDiabloMsg(EMSG_SHRINE_RELIGIOUS);
 }
 
-void OperateShrineEnchanted(Player &player)
+void OperateShrineEnchanted(Player &player, uint32_t seed)
 {
 	if (&player != MyPlayer)
 		return;
 
-	int cnt = 0;
-	uint64_t spell = 1;
-	int maxSpells = gbIsHellfire ? MAX_SPELLS : 37;
-	uint64_t spells = player._pMemSpells;
-	for (int j = 0; j < maxSpells; j++) {
-		if ((spell & spells) != 0)
-			cnt++;
-		spell *= 2;
-	}
-	if (cnt > 1) {
-		spell = 1;
-		for (int j = SPL_FIREBOLT; j < maxSpells; j++) { // BUGFIX: < MAX_SPELLS, there is no spell with MAX_SPELLS index (fixed)
-			if ((player._pMemSpells & spell) != 0) {
-				if (player._pSplLvl[j] < MaxSpellLevel)
-					player._pSplLvl[j]++;
+	const size_t lastSpellIndex = static_cast<size_t>(gbIsHellfire ? spell_id::SPL_LAST : spell_id::SPL_LASTDIABLO);
+
+	std::bitset<MAX_SPELLS> knownSpells { player._pMemSpells & (~0ULL >> (64 - lastSpellIndex)) };
+	// If the player knows any spells
+	if (knownSpells.any()) {
+		// choose a spell at random to decrement (+1 as spells are 1-indexed in Player::_pSplLvl)
+		unsigned r = ShiftModDistribution(seed, knownSpells.count()) + 1;
+
+		// r currently represents the nth set bit, translate this to the actual bit position
+		for (size_t i = 0; i < r; i++) {
+			if (!knownSpells.test(i))
+				r++;
+		}
+
+		for (size_t j = static_cast<size_t>(SPL_FIREBOLT); j <= lastSpellIndex; j++) {
+			if ((player._pMemSpells & GetSpellBitmask(static_cast<spell_id>(j))) != 0) {

```

Comment:
This is pretty confusing. The first for loop isn't really getting the actual bit position, but rather the 1-based spell index. Also, is there some reason we shouldn't use the same approach in both for loops (`std::bitset` vs `GetSpellBitmask()`)?

Reply:
we could make player._pMemSpells a bitset, it'd require a bit more refactoring to update the uses and save/load of that field as well though.

Regarding the 0-based/1-based confusion, player._pMemSpells is 0-based, but player._pSplLvl is 1-based as player._pSplLevel[0] corresponds to the class skill (and is always 0). It could be rewritten like the following to make it clearer:
```cpp
size_t r = ShiftModDistribution(knownSpells.count());

for (size_t i = 0; i < r; i++) {
    if (!knownSpells.test(i))
        r++;
}

r++; // translate from 0-based index as used in _pMemSpells to 1-based as used for _pSplLvl
```

Could also just use a single loop and use r as a decrementing counter of how many known spells to skip.

Reply:
I didn't necessarily mean to suggest we make `player._pMemSpells` a bitset, although maybe that would be an overall improvement. I'm really just talking about how the bitset is literally an equivalent representation of `player._pMemSpells` within the scope of this particular PR. Consider this implementation of the first loop.

```c++
for (size_t i = static_cast<size_t>(SPL_FIREBOLT); i <= r; i++) {
	if ((player._pMemSpells & GetSpellBitmask(static_cast<spell_id>(i))) == 0)
		r++;
}
```

One could similarly reimagine the second loop using the bitset instead of `GetSpellBitmask()`. IMO, either way you slice it, it doesn't make sense to use a bitset for the first loop and then `GetSpellBitmask()` for the second loop when they essentially do the same thing.

Reply:
The bitset currently is pretty much only used to get the count of known spells for the currently active game mode. In most cases it'll be equivalent to `Player::_pMemSpells` however I wasn't sure whether loading a hellfire character into diablo masked off the hellfire specific spells on load (and didn't bother to check...). The bitmask is constructed by ANDing with a bitmask representing all valid spells for the given game mode as a safety measure. I should comment that better, but I'll have a go at refactoring `_pMemSpells` when I get time and revise this PR instead.

In the current state of this PR the first loop is primarily concerned with the bitset of known spells within the given game mode, which may not be the same as `Player::_pMemSpells`.

The second loop uses `spell_id` enum values as indexes into `Player::_pSplLvl` as the primary interaction. I kept it using `GetSpellBitmask()` to document the translation from `spell_id` as an index to the appropriate `Player::_pMemSpells` bit position.

Reply:
Upon reflection, I think the bitset itself is more of a minor problem compared to the sleight-of-hand pulled with the variable `r`. Ultimately, I just feel this code should just be written in a way that is more conceptually accurate.

* Your first comment claims we add 1 because the we need to use a 1-based index for `player._pSplLvl`. The very next comment immediately reinterprets it as the nth known spell, which I think is the more accurate interpretation based on how it's generated.
* `r` is then used as the upper bound for a for loop that uses a 0-based index for the bitset. This makes the least amount of sense because `i` doesn't really represent the ith known spell. This feels more like a micro-optimization made possible by the dual interpretation of the variable `r`.
* When the loop is complete, `r` somewhat magically points to the `spell_id` of the nth known spell. Conceptually, the loop is shifting `r` over by the number of unknown spells with a smaller index than the nth known spell, retroactively suggesting maybe the first comment was right all along.